Alignment/CommonAlignmentAlgorithm/src/AlignmentParameterStore.cc

```diff
@@ -445,53 +446,89 @@ applyAlignableAbsolutePositions( const align::Alignables& alivec,
-	  LogDebug("NewPosition") << "moving by:" << posDiff;
-	  LogDebug("NewRotation") << "rotating by:\n" << rotDiff;
-
-	  // add position error
-	  // AlignmentPositionError ape(shift.x(),shift.y(),shift.z());
-	  // (*iali)->addAlignmentPositionError(ape);
-	  // (*iali)->addAlignmentPositionErrorFromRotation(rot);
-				
-	  found=true;
-	  ++nappl;
-	}
+        if (found) {
+          edm::LogError("DuplicatePosition")
+            << "New positions for alignable found more than once!";
+        }
+        else {
+          // New position/rotation
+          const align::PositionType& pnew = ipos->pos();
+          const align::RotationType& rnew = ipos->rot();
+          const std::vector<double>& dnew = ipos->deformationParameters();
+          // Current position / rotation
+          const align::PositionType& pold = ali->globalPosition();
+          const align::RotationType& rold = ali->globalRotation();
+          // Current surf. deformation
+          std::vector<std::pair<int, SurfaceDeformation*> > dold_id_pairs;
+          SurfaceDeformation* dold_obj=0;
+          SurfaceDeformationFactory::Type dtype = SurfaceDeformationFactory::kNoDeformations;
+          std::vector<double> dold;
+          if (1 == ali->surfaceDeformationIdPairs(dold_id_pairs)){ // might not have any...
+            dold_obj = dold_id_pairs[0].second;
+            dold = dold_obj->parameters();
+            dtype = (SurfaceDeformationFactory::Type)dold_obj->type();
+          }
+
+          // shift needed to move from current to new position
+          align::GlobalVector posDiff = pnew - pold;
+          align::RotationType rotDiff = rold.multiplyInverse(rnew);
+          align::rectify(rotDiff); // correct for rounding errors 
+          ali->move(posDiff);
+          ali->rotateInGlobalFrame(rotDiff);
+          LogDebug("NewPosition") << "moving by:" << posDiff;
+          LogDebug("NewRotation") << "rotating by:\n" << rotDiff;
+
+          // add the surface deformations
+          // If an old surface deformation record exists, ensure that the added deformation has the same type and size.
+          if (!dold.empty() && dtype != SurfaceDeformationFactory::kNoDeformations && dnew.size()==dold.size()){
+            std::vector<double> defDiff; defDiff.reserve(dold.size());
+            for (unsigned int i = 0; i < dold.size(); i++) defDiff.push_back(dnew[i] - dold[i]);
+            auto deform = SurfaceDeformationFactory::create(dtype, defDiff);
+            edm::LogInfo("Alignment") << "@SUB=AlignmentParameterStore::applyAlignableAbsolutePositions"
+              << "Adding surface deformation of type " << SurfaceDeformationFactory::surfaceDeformationTypeName((SurfaceDeformationFactory::Type)deform->type())
+              << ", size " << defDiff.size() << " and first element " << defDiff.at(0)
+              << " to alignable with id / type: " << id << " / " << typeId;
+            ali->addSurfaceDeformation(deform, true);
```

Comment:
would be better than ali took ownership of this pointer so that it didn't have to make a copy of it - but that looks like an old problem...